### ISSUE TITLE
compilation failed with the new cython version (0.28)

### ISSUE THREAD

**Issue by skoudoro** [nipy/dipy#1435]
## Description

As you can see on [this build](https://travis-ci.org/nipy/dipy/jobs/341509214#L1437), the new Cython version (0.28) does not allow to use memoryviews as struct members.

In [this discussion](https://github.com/cython/cython/issues/1453), it is hard to say if it will be something permanent or not.

BTW, We have to find a workaround.

**Comment by MarcCote** [comment]
If I remember correctly, this comment (https://github.com/cython/cython/issues/1453#issuecomment-364864228) is exactly what we are doing, i.e. manually nulling the memview at creation and freeing it in dealloc.

**Comment by Garyfallidis** [comment]
Check the travis error the message is clear. Memory views not allowed as struct members.

**Comment by Garyfallidis** [comment]
And even if you nullify it the possibility for memory leak is still there.

**Comment by Garyfallidis** [comment]
Basically, they are not allowing the code to compile when we add memviews inside structs. We can suggest to reconsider not changing this option.

**Comment by Garyfallidis** [comment]
```
cdef struct Centroid:
    Data2D features # float* features
    int size
    float[6] aabb
```

**Comment by Garyfallidis** [comment]
Perhaps changing from Data2D to float * will do the job? Let us know if you have time to hangout on this. We need to coordinate.

**Comment by Garyfallidis** [comment]
Otherwise perhaps we can change from struct to class for structures with memviews?

**Comment by Garyfallidis** [comment]
Okay @MarcCote the class idea was not good. 
But I think we can make something like this

```
cdef struct Centroid2:
    float * features #using pointer rather than memview
    int size
```
and then have functions to move from pointer to memview and memview to pointer
```
def memview_to_pointer(float [:, ::1] mv):
    cdef Centroid2 c
    c.features = <float *> &mv[0, 0]

def pointer_to_memview(size_t N, size_t M):

    cdef Centroid2 c
    c.features = <float *>PyMem_Malloc(N * M * sizeof(float))  
    cdef float[:, ::1] mv = <float[:N, :M]> c.features
    
def free_memview(float [:, ::1] mv): #not working
    if mv._data is not NULL:
        PyMem_Free(mv._data) 
``` 
When freeing the memory the IPython kernel is restarting. Let us know if you know a better way to free the memview.

**Comment by MarcCote** [comment]
I like the Centroid2 struct idea. Do we really need the conversion functions, though?

**Comment by Garyfallidis** [comment]
The data are coming in as memviews or numpy arrays. And you use 2D memviews internally too. So... probably yes. But if you think otherwise we would love to hear an alternative.

**Comment by Garyfallidis** [comment]
We should take care this issue because it is holding up release 0.14. So, it is high priority! :)

**Comment by MarcCote** [comment]
I don't have a precise idea on how to do it. I wasn't able to install cython 0.28 on my machine. Is 0.28 officially released?

**Comment by skoudoro** [comment]
This is a future release @MarcCote.  Create a new environment and you can install it via : 

```
pip install --find-links=https://5cf40426d9f06eb7461d-6fe47d9331aba7cd62fc36c7196769e4.ssl.cf2.rackcdn.com --find-links=https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com --pre cython numpy scipy h5py nibabel
```

**Comment by skoudoro** [comment]
fix by #1434, closing !
